third-party/dparser/d/make_dparser.1
.TH make_dparser 1 "Mar 2003" Unix "User Manuals"
.SH NAME
.PP
make_dparser \- D Parser Generator
.SH SYNOPSIS
.PP
.B make_dparser
[\fIoptions\fP] \fIgrammar\fP
.SH DESCRIPTION
.PP
make_dparser is a scannerless GLR parser generator based on the Tomita
algorithm. It is self-hosted and very easy to use. Grammars are written in a
natural style of EBNF and regular expressions and support both speculative and
final actions.
.SH OPTIONS
.PP
.IP "-l"
Toggle longest match lexical ambiguity resolution.
When ON the scanner only recognizing the longest matching tokens
in a given state. This provides an easy way to build
grammars for languages which use longest match lexical
ambiguity resolution (e.g. ANSI-C, C++). (OFF by default)
.IP "-T"
Toggle building of a tokenizer for START.  When ON, instead of generating 
a unique scanner for each state (i.e. a 'scannerless' parser), 
generate a single scanner (tokenizer) for the entire grammar.  This provides
an easy way to build grammars for languages which assume a
tokenizer (e.g. ANSI C). (OFF by default)
.IP "-H \fI-1/0/1\fP"
Write header, 0 : no, 1 : yes, -1 : only if not empty.
.IP "-t \fI0/1\fP"
Token type, 0 : #define, 1 : enum.
.IP "-C"
Toggle computing whitespace states.  If 'whitespace' is
defined in the grammar, then use it as a subparser to 
consume whitespace. (ON by default)
.IP "-A"
Toggle computing states for all non-terminals.  Ensures that there is a unique
state for each non-terminal so that a subparsers can be invoked for
that non-terminal. (OFF by default)
.IP "-i \fItag\fP"
Tag for grammar data structures so that multiple sets of tables can
be included in one file/application. (defaults to 'gram')
.IP "-b \fIblocks\fP"
Number of blocks to which scanner tables are broken up into.
Larger numbers permit more sharing with more overhead.  4 seems to be optimal
for most grammars. (defaults to 4)
files.
.IP "-p"
Toggle setting of operator priority from rules.  Setting of operator
priorities on operator tokens can increase the size of the tables but 
can permit unnecessary parse stacks to be pruned earlier. (OFF by default)
.IP "-r"
Toggle use of right recursion for EBNF productions.  Do not change this
unless you really know what you are doing. (OFF by default) 
.IP "-L"
Toggle writing of line numbers.  Used to debug the parsing table
generator itself. (ON by default)
.IP "-X"
Set the extension of the generated code file.  For C++ programs (for example)
the extension can be set to .cpp with the option "-Xcpp".  ("-Xc" by default)
.IP "-v"
Increase verbosity.
.IP "-d"
Increase debugging.
.IP "-h"
Display help.
.SH ENVIRONMENT
.PP
All the options are available through environment variables
with the names:

.na
.nh
D_MAKE_PARSER_HEADER,
D_MAKE_PARSER_TOKEN,
D_MAKE_PARSER_GRAMMAR_IDENT,
D_MAKE_PARSER_SCANNER_BLOCKS,
D_MAKE_PARSER_SET_PRIORITY,
D_MAKE_PARSER_WRITE_LINE_DIRECTIVES,
D_MAKE_PARSER_VERBOSE,
D_MAKE_PARSER_DEBUG.
.ad
.hy
.SH FEATURES
.PP
The features are covered in the documentation.  See the README file.
.SH FILES
.PP
None.	
.SH NO WARRANTIES
This program is distributed in the hope that it will be useful,
but WITHOUT ANY WARRANTY; without even the implied warranty of
MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
COPYRIGHT for more details.
.SH SEE ALSO
.PP
.BR flex (1),
.BR yacc (1),
.BR bison (1)
.PP
D Parser Home Page: http://dparser.sourceforge.net/
.PP
The BSD General Public License.
.SH AUTHOR
.PP
John Plevyak.  Use <jplevyak at users.sourceforge.net> to contact.



CHANGES
====================
Release Changes List
====================

finished thru commit 5d9bf0e

==============
version 1.13.0
==============

Sixteenth public release of Chapel, April 7, 2016

Highlights
----------
TODO

Configuration Changes
---------------------
* made 'jemalloc' our default allocator for all platforms other than Cygwin
* made 'cray-prgenv-cray' default to using qthreads+hwloc and 'ugni'
* added support for 're2' when using Cray and PGI compilers
* changed 'knc' to default to using 'qthreads+hwloc'
* switched to using pg++ rather than pgCC as the C++ compiler for 'pgi'
* removed stale support for Xcode

Semantic Changes / Changes to Chapel Language
---------------------------------------------
* made default I/O routines ignore 'param' fields as they have 'type' fields
* improved the handling of "return by ref" functions in setter/getter contexts
  (see 'The Ref Return Intent' under 'Procedures' in the language spec)
* restricted 'sync'/'single' types to sufficiently simple types
* removed support for performing Chapel I/O on 'c_string's
* removed coercions from 'c_string's to 'string's

New Features
------------
* added a standard library of routines on the 'string' type
  (see http://chapel.cray.com/docs/1.13/modules/internal/String.html)
* added support for filtering module 'use' statements with 'only' and 'except'
  (e.g., 'use M only foo;' or '... except foo;' to restrict to / exclude 'foo')
  (see 'Using Modules' in spec)
* added support for renaming 'use'd module symbols
  (e.g., 'use M only foo as M_foo;' causes 'foo' to be referred to as 'M_foo')
  (see 'Using Modules' in the spec)
* added the ability to 'use' an enumerated type
  (e.g., "...mycolor.green..." can now be written "use mycolor;  ...green...")
* added support for casting types to strings and deprecated typeToString()
  (e.g., 'writeln(x.type:string);' can be used to print the name of x's type)
* added a capability to iterate over the fields in a class or record
  (see http://chapel.cray.com/docs/1.13/modules/standard/Reflection.html)
* added the ability to query whether a call can be resolved or not
  (see http://chapel.cray.com/docs/1.13/modules/standard/Reflection.html)
* added support for annotating an on-clause as being 'local'
  (see http://chapel.cray.com/docs/1.13/technotes/local.html)
* added support for 'c_string' values
  (e.g., "this is a Chapel string" while c"this is a C string")

Feature Improvements
--------------------
* added support for creating subclasses of a generic class
* added support for min/max reduce intents
* added support for reduce intents on coforall loops
* added the ability to pass sync/single variables via generic arguments
* improved the type signatures of min()/max() to support better type resolution
* extended comparison operators (==, !=, <=, etc.) to support mixed int/uints
* added support for range slicing between mixed int/uint idxTypes
* added support for comparing between 'c_void_ptr' expressions and 'nil'
* added support for implicitly coercing 'c_ptr' types to 'c_void_ptr'
* added support for casting from fully-qualified enum strings to enum values
  (e.g., '"mycolor.green":mycolor' now works where it wouldn't have previously)
* extended channel.readbits/writebits to accept other integral types
* added scalar versions of domain.exterior(), interior(), and translate()

Standard Library/Modules
------------------------
* started distinguishing between "standard Chapel modules" and "packages"
  (see http://chapel.cray.com/docs/1.13/modules/modules.html
   and http://chapel.cray.com/docs/1.13/modules/packages.html)
* added a new JAMA-based matrix package
  (see TODO)
* added a new PCG-based random number generator to 'Random' as the default
  (see http://chapel.cray.com/docs/1.13/modules/standard/Random/PCGRandom.html)
* added a new 'Reflection' module supporting reflection about a program
  (see http://chapel.cray.com/docs/1.13/modules/standard/Reflection.html)
* extended the 'HDFS' module to support libhdfs3
  (http://chapel.cray.com/docs/1.13/technotes/auxIO.html)
* added support for skipping unknown fields when reading records/classes
  (see TODO))
* made the default I/O routines for records/classes support arbitrary orders
* added 'basename', 'dirname', and 'splitPath' to the 'Path' module
  (see http://chapel.cray.com/docs/1.13/modules/standard/Path.html)
* added rmTree(), moveDir() to the 'FileSystem' module
  (see http://chapel.cray.com/docs/1.13/modules/standard/FileSystem.html)
* added an optional time units argument to 'sleep()' in the Time module
  (see http://chapel.cray.com/docs/1.13/modules/standard/Time.html#Time.sleep)
* added support for sending signals to subprocesses for the 'Spawn' module
  (see http://chapel.cray.com/docs/1.13/modules/standard/Spawn.html)
* added an optional 'buffer' boolean to the 'Spawn' module's wait() routine
  (see http://chapel.cray.com/docs/1.13/modules/standard/Spawn.html#Spawn.subprocess.wait)
* added an isclosed() method to the 'channel' type
  (see http://chapel.cray.com/docs/1.13/modules/standard/IO.html#IO.channel.isclosed)
* improved the 'List' module to support JSON format
* improved error handling in the IO and Spawn modules
* added the ability to call exit() without arguments for a 0 default status
* removed deprecated functions from 'Memory' module

Performance Improvements
------------------------
* used an amortized reallocation strategy for resizing arrays due to vector ops
* made bulk data transfers more likely to fire by removing a conservative check
* enabled bulk data transfers between portions of the same array
* generally improved the performance of string-based computations
* optimized iteration over anonymous range values of the form 'lo..#size'
* improved the performance of code generated by the LLVM back-end
* improved the performance of many common reduction cases

Memory Improvements
-------------------
* dramatically reduced the compiler-generated leaks due to strings and records

third-party/dparser/d/make_dparser.c
/*
  Copyright 2002-2004 John Plevyak, All Rights Reserved
*/
#include "gramgram.h"
#include "d.h"

static void help(ArgumentState *arg_state, char *arg_unused);

static int set_op_priority_from_rule = 0;
static int right_recursive_BNF = 0;
static int states_for_whitespace = 1;
static int states_for_all_nterms = 0;
static int tokenizer = 0;
static int longest_match = 0;
static char grammar_ident[256] = "gram";
static int ident_from_filename = 0;
static int scanner_blocks = 4;
static int scanner_block_size;
static int write_line_directives = 1;
static int write_header = -1;
static int token_type = 0;
static char write_extension[256] = "c";

static ArgumentDescription arg_desc[] = {
 {"longest_match", 'l', "Use Longest Match Rule for Tokens", "T", 
  &longest_match, "D_MAKE_LONGEST_MATCH", NULL},
 {"tokenizer", 'T', "Tokenizer for START", "T", &tokenizer,
  "D_MAKE_PARSER_TOKENIZER", NULL},
 {"write_header", 'H', "Write Header (-1:if not empty)", "I", &write_header,
  "D_MAKE_PARSER_HEADER", NULL},
 {"token_type", 't', "Token Type (0:define, 1:enum)", "I", &token_type,
  "D_MAKE_PARSER_TOKEN", NULL},
 {"whitespace_states", 'C', "Compute Whitespace States", "T", 
  &states_for_whitespace,  "D_MAKE_PARSER_WHITESPACE", NULL},
 {"all_states", 'A', "Compute States For All NTERMs", "T", 
  &states_for_all_nterms,  "D_MAKE_PARSER_ALL_NTERMS", NULL},
 {"grammar_ident", 'i', "Grammar Identifier", "S255", grammar_ident,
  "D_MAKE_PARSER_GRAMMAR_IDENT", NULL},
 {"filename_ident", 'I', "Use Filename as Identifier", "T", &ident_from_filename,
  "D_MAKE_PARSER_IDENT_FROM_FILENAME", NULL},
 {"scanner_blocks", 'b', "Scanner Blocks", "I", &scanner_blocks,
  "D_MAKE_PARSER_SCANNER_BLOCKS", NULL},
 {"op_pri_from_rule", 'p', "Set Operator Priority From Rule", "T", 
  &set_op_priority_from_rule, "D_MAKE_PARSER_SET_PRIORITY", NULL},
 {"right_recurse_BNF", 'r', "Use Right Recursion For */+", "T", 
  &right_recursive_BNF, "D_MAKE_PARSER_RIGHT_RECURSIVE_BNF", NULL},
 {"write_lines", 'L', "Write #line(s)", "T", 
  &write_line_directives, "D_MAKE_PARSER_WRITE_LINE_DIRECTIVES", NULL},
 {"ext", 'X', "Code file extension (e.g. cpp)", "S256", 
  &write_extension[0], "D_MAKE_PARSER_WRITE_EXT", NULL},
 {"verbose", 'v', "Verbose", "+", &d_verbose_level, 
  "D_MAKE_PARSER_VERBOSE", NULL},
 {"debug", 'd', "Debug", "+", &d_debug_level, 
  "D_MAKE_PARSER_DEBUG", NULL},
 {"help", 'h', "Help", NULL, NULL, NULL, help},
 {0}
};

static ArgumentState arg_state = {
  0, 0,
  "program", 
  arg_desc
};

extern D_ParserTables parser_tables_dparser_gram;

static void
help(ArgumentState *arg_state, char *arg_unused) {
  char ver[30];
  d_version(ver);
  fprintf(stderr, "Make DParser Version %s ", ver);  
  fprintf(stderr, "Copyright (c) 2002-2004 John Plevyak\n");  
  usage(arg_state, arg_unused);
}

int
main(int argc, char *argv[]) {
  char *grammar_pathname;
  Grammar *g;

  process_args(&arg_state, argv);
  if (arg_state.nfile_arguments != 1)
    help(&arg_state, NULL);
  grammar_pathname = arg_state.file_argument[0];
  g = new_D_Grammar(grammar_pathname);
  /* grammar construction options */
  g->set_op_priority_from_rule = set_op_priority_from_rule;
  g->right_recursive_BNF = right_recursive_BNF;
  g->states_for_whitespace = states_for_whitespace;
  g->states_for_all_nterms = states_for_all_nterms;
  g->tokenizer = tokenizer;
  g->longest_match = longest_match;
  /* grammar writing options */
  strcpy(g->grammar_ident, grammar_ident);
  if (ident_from_filename) {
    char *n = strrchr(grammar_pathname, '/'), *e;
    n = n ? n : grammar_pathname;
    e = strchr(n, '.');
    e = e ? e : n + strlen(n);
    memcpy(g->grammar_ident, n, e-n);
    g->grammar_ident[e-n] = 0;
  }
  g->scanner_blocks = scanner_blocks;
  g->scanner_block_size = scanner_block_size;
  g->write_line_directives = write_line_directives;
  g->write_header = write_header;
  g->token_type = token_type;
  strcpy(g->write_extension, write_extension);

  if (parse_grammar(g, grammar_pathname, &parser_tables_dparser_gram, sizeof(ParseNode_User)) < 0)
    d_fail("unable to parse grammar '%s'", grammar_pathname);
  if (g->productions.n < 2)
    d_fail("no productions in grammar '%s'", grammar_pathname);
  if (build_grammar(g) < 0)
    d_fail("unable to load grammar '%s'", grammar_pathname);
  if (write_ctables(g) < 0)
    d_fail("unable to write C tables '%s'", grammar_pathname);
  free_D_Grammar(g);
  return 0;
}


